### ISSUE TITLE
Add support for Buildpack API 0.6

### ISSUE THREAD

**Issue by edmorley** [Malax/libcnb.rs#95]
Currently after #50, libcnb enforces that the buildpack is using buildpack API 0.5:
https://github.com/Malax/libcnb.rs/blob/3180fb553fce45c70740c06eb9b13a177493e486/src/lib.rs#L63

The latest buildpack API version is 0.6 - we should add support for the changes included in that, and then bump libcnb's expected/supported version.

See:
https://github.com/buildpacks/spec/releases/tag/buildpack%2Fv0.6

First step for this issue is to identify which of the changes in that release affect libcnb, and to document them here.

Cross-reference [GUS-W-9625362](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07AH000000RebNYAS/view).

**Comment by schneems** [comment]
I created a 0.6.0 milestone. Looking at the changelog here's my understanding:

## Additions

- Buildpacks may contribute the default process type (#159, #205, RFC 0064).
  - Implemented by https://github.com/Malax/libcnb.rs/pull/94
- Adds exec.d support on Windows (#164, #203).
  - I don't know what this needs I see https://github.com/Malax/libcnb.rs/blob/main/src/data/exec_d.rs but I don't know what this feature does or what the code does. Looks like exec.d is a place to put executables without having to modify the path?
  - Since I don't know what this feature does I would say it might conflict, I need a second set of eyes.
- Adds new descriptor keys to buildpack.toml (#181, #208, RFC 0070).
  - These are all optional:
	- Add description as a free text field
	- Add licenses table
	- Add keywords array
  - I wanted to see if buildpack.rs blows up with these changes, it does not:

```
    fn can_serialize_minimal_buildpack() {
        let raw = r#"
api = "0.4"

[buildpack]
id = "foo/bar"
name = "Bar Buildpack"
version = "0.0.1"
keywords = [ "<string>" ]

[[buildpack.licenses]]
type = "<string>"
uri = "<uri>"

[[stacks]]
id = "io.buildpacks.stacks.bionic"

[metadata]
checksum = "awesome"
"#;

        let result = toml::from_str::<BuildpackToml<toml::value::Table>>(raw);
        assert!(result.is_ok());
        if let Ok(toml) = result {
            assert!(!toml.buildpack.clear_env);
        }
    }
```

```
$ cargo test buildpack
    Finished test [unoptimized + debuginfo] target(s) in 0.03s
     Running unittests (target/debug/deps/libcnb-ef65c55f38f8d2d3)

running 11 tests
test data::buildpack::tests::buildpack_id_does_not_allow_app ... ok
test data::buildpack::tests::buildpack_id_does_not_allow_config ... ok
test data::buildpack::tests::buildpack_api_display ... ok
test data::buildpack_plan::tests::it_parses_empty ... ok
test data::buildpack_plan::tests::it_parses_simple ... ok
test data::buildpack_plan::tests::it_parses_with_metadata ... ok
test data::buildpack::tests::buildpack_id_from_str_major ... ok
test data::buildpack::tests::buildpack_id_from_str_major_minor ... ok
test data::buildpack::tests::can_serialize_minimal_metabuildpack ... ok
test data::buildpack::tests::can_serialize_minimal_buildpack ... ok
test data::buildpack::tests::can_serialize_metabuildpack ... ok

test result: ok. 11 passed; 0 failed; 0 ignored; 0 measured; 16 filtered out; finished in 0.00s
```


## Breaking Changes

- Buildpacks must explicitly opt-in to layer re-use by appending to <layers>/<layer>.toml
  - If buildpacks want caching, they have to write `cache = true` to their `<layer>.toml` on every build versus previously that toml would be restored and ther value will persist even if the buildpack author doesn't write to that value.
  - It sounds like for us it means we always write the full value to a `<layer>.toml`. Looking at LayerContentMetadata I think we're always serializing all values

- Adds types table to <layers>/<layer>.toml.
  - Adds an extra optional way of specifying layer keys (launch, build, cache) at the bottom of the file for bash users.
  - If we want to work with toml files written by other buildpacks as opposed to only toml files written by us, then we need to implement this logic. For now we're only writing and reading our own toml files, I think we should file this as a must do but also shouldn't be a blocker for rev-ing the supported version to 0.6

## Nonfunctional Changes

- Explicitly bans the use of special layer names by buildpacks (#202)
  - We could detect and error on these names, but I think pack will do that too. Probably a good idea for feature-complete spec implementation, but not strictly necessary to bump the minimum version to 0.6
- Clarifies the role of the launcher in setting environment variables at runtime (#207)
  - Text: "For each file written to `<layers>/<layer>/env/` or `<layers>/<layer>/env.launch/` by `/bin/build`, the lifecycle MUST modify an environment variable during the launch phase according to the modification rules below (see [launcher](platform.md#launcher))."
  - I don't think this affects anything, our code doesn't run at launch time
- Fixes links to exec.d interface (#168)
  - This was contributed in december 2020 by our dear own Manuel 🎉

**Comment by Malax** [comment]
This is great, both the milestone itself and the changelog digging! Thanks! 👍🏻

**Comment by schneems** [comment]
I think I mis-read the breaking change around `[types]` on the first read through i.e. "Adds types table to /.toml." It looks like it changes:

```
launch = false
build = false
cache = false
```

to 

```
[types]
launch = false
build = false
cache = false
```

And doesn't seem to be optional (despite the giant words "optional" in the RFC): https://github.com/buildpacks/spec/pull/209/files. To support this we need to update either the LayerContentMetadata or at least how it's serialized. I'm not sure if there's a "clean" solution that doesn't require breaking existing code.

```rs
pub struct LayerContentMetadata<M> {
    /// Whether the layer is intended for launch.
    #[serde(default = "defaults::r#false")]
    pub launch: bool,

    /// Whether the layer is intended for build.
    #[serde(default = "defaults::r#false")]
    pub build: bool,

    /// Whether the layer is cached.
    #[serde(default = "defaults::r#false")]
    pub cache: bool,

    /// Metadata that describes the layer contents.
    pub metadata: M,
}
```

**Comment by edmorley** [comment]
I've confirmed we don't need to do anything for this one:
> Adds exec.d support on Windows (https://github.com/buildpacks/spec/issues/164, https://github.com/buildpacks/spec/pull/203).

...since (a) the changes were only for Windows, (b) the change only involved the expected inputs/outputs of `exec.d/` binaries, and libcnb.rs doesn't have any support for creating those binaries at the moment.

**Comment by edmorley** [comment]
Confirmed these need no changes on our side...

> - Clarifies the role of the launcher in setting environment variables at runtime (https://github.com/buildpacks/spec/pull/207)
...
> - Fixes links to exec.d interface (https://github.com/buildpacks/spec/pull/168)

These are just spec wording clarifications/fixes, no actual change.

**Comment by edmorley** [comment]
> Explicitly bans the use of special layer names by buildpacks (https://github.com/buildpacks/spec/pull/202) 

I've filed #101, however did not put it in the Buildpack API 0.6 milestone, since it's more of a nice-to-have rather the something we need for 0.6 support.

**Comment by edmorley** [comment]
@hone I checked against buildpack API 0.5 too, and found #104 (not implemented), and #105 (implemented, but not correctly).
